ls1intum/Artemis
# `Discussion`: Improve usability in discussion overview

### Checklist
#### General


#### Server
#### Client
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the test guidelines.

### Motivation and Context


Some things were on my To-Do-List:
- Remove `votes` from the `Post` class, as the data model changed and voting is reflected by `reactions` (the data column will be dropped in a follow-up to ensure compatibility of DB's on test servers)

- Renaming: `PostContentCompareStrategy`-> `PostSimilarityComparisonStrategy`, as the check does not necessarily depends on the content but also take other `Post` Metadata in account when checking for similarity
- Renaming: `PageDiscussionSectionComponent` -> `DiscussionSectionComponent`, no need to have `Page` in the naming
- Renaming: all components that include both types, `Post` and `AnswerPost` had `Posting` (superclass) in the name, but with inconsistent convention `postings-xy` vs. `posting-xy`, now changed to `posting-xy`

- Discussion Overview: options that can be selected as context, i.e., lectures and exercises are sorted alphabetically to enable users to quickly find the context they want to select
- Discussion Overview: when a user started to create a post, cancelled it (clicking x) and then opened the modal again, the context selection was not properly reset (first input field of the modal), it should be `Tech Support` on default
- Discussion Overview: Announcement (pinned and highlighted posts) were not listed on top of the list if there are other posts that were created afterwards and also pinned, because the creation date was the second criterion to order posts after the display priority criterion (pinned, archived). I introduced the announcement type criterion as first order criterion now. 

### Description

All those changes are not complex to understand, most important is, that it does not breaking a working UI.
The tests I had to touch were also updated in accordance with our newest client test guidelines.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Course with posts enabled

1. Log in to Artemis
2. Navigate to discussion overview
3. As instructor start to create an announcement (select Announcement as context), and cancel before saving, i.e. click `x` at the top right of the modal
4. Start to create a new post again by clicking the button `New Post` -> check that the context is set to Tech Support on start-up
5. Select announcement as context -> check if the blue info occurs right below the context
6. Select any other context and provide a title -> check that similar posts are shown below in a list
7. Save the post
8. Interact with existing posts -> editing, deleting, react, answering -> check that no unexpected behavior occurs
9. Create an announcement and another post afterwards. Pin the other post -> the announcement should be listed at top (even though the second post was created more recently and is also pinned)
10. Remove the pin of the announcement -> it should be listed after it's creation date
11. Archive the announcement -> it should be moved to the bottom of the list
12. Log in as student and interact with a posting -> create, edit own, delete own, answer -> check that no unexpected behavior occurs

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage






### Screenshots



Instructor who is checking the contexts which are sorted alphabetically and cancels + creates a post as described in the testing steps.

https://user-images.githubusercontent.com/41366522/142689015-85af9d79-b452-4cfd-9704-31fc82ea1b63.mov
